tty/pc-tty/ttypc_kbd.c

```diff
@@ -402,165 +417,205 @@ static char *_ttypc_kbd_get(ttypc_t *ttypc)
+	else if (!strcmp(s, "\033[B") && ((ttypc->lockst & KB_SCROLL) || (ttypc->shiftst == (KB_CTL | KB_SHIFT)))) {
+		_ttypc_vga_scroll(cvt, -1);
+	}
+	/* Scroll up one page */
+	else if (!strcmp(s, "\033[5~") && ((ttypc->lockst & KB_SCROLL) || (ttypc->shiftst == KB_SHIFT))) {
+		_ttypc_vga_scroll(cvt, cvt->rows);
+	}
+	/* Scroll down one page */
+	else if (!strcmp(s, "\033[6~") && ((ttypc->lockst & KB_SCROLL) || (ttypc->shiftst == KB_SHIFT))) {
+		_ttypc_vga_scroll(cvt, -cvt->rows);
+	}
+	/* Scroll to top */
+	else if (!strcmp(s, "\033[H") && ((ttypc->lockst & KB_SCROLL) || (ttypc->shiftst == KB_SHIFT))) {
+		_ttypc_vga_scroll(cvt, cvt->scrbsz - cvt->scrbpos);
+	}
+	/* Scroll to bottom */
+	else if (!strcmp(s, "\033[F") && ((ttypc->lockst & KB_SCROLL) || (ttypc->shiftst == KB_SHIFT))) {
+		_ttypc_vga_scroll(cvt, -cvt->scrbpos);
+	}
+	/* Switch between VTs (up to 12) */
+	else if (!strncmp(s, "\033[", 2) && (s[2] >= 'm') && (s[2] <= 'x') && !s[3]) {
+		if ((s[2] - 'm') < NVTS)
+			_ttypc_vga_switch(ttypc->vts + (s[2] - 'm'));
+	}
+	/* Regular character processing */
+	else {
+		/* Process KP/FN keys */
+		if (!strncmp(s, "\033[", 2)) {
+			m = kpmod[ttypc->shiftst & 0x0f];
+
+			if ((k = s[2])) {
+				if (!s[3] && ((k >= 'A' && k <= 'F') || k == 'H')) {
+					if (m) {
+						snprintf(buff, sizeof(buff), "\033[1;%u%c", m, k);
+						s = buff;
 					}
-					else if ((s[3] == '~') && !s[4] && (k >= '2' && k <= '6' && (k != '4'))) {
-						if (m) {
-							snprintf(buff, sizeof(buff), "\033[%c;%u~", k, m);
-							s = buff;
-						}
+					/* Cursor Key Mode */
+					else if (cvt->ckm) {
+						snprintf(buff, sizeof(buff), "\033O%c", k);
+						s = buff;
+					}
+				}
+				else if ((s[3] == '~') && !s[4] && (k >= '2' && k <= '6' && (k != '4'))) {
+					if (m) {
+						snprintf(buff, sizeof(buff), "\033[%c;%u~", k, m);
+						s = buff;
 					}
 				}
 			}
-			else if (!strcmp(s, "\r") || !strcmp(s, "\n")) {
-				/* Line feed/New line Mode */
-				if (cvt->lnm)
-					s = "\r\n";
-			}
-
-			while (*s && !libtty_putchar(&cvt->tty, *s++, NULL));
+		}
+		else if (!strcmp(s, "\r") || !strcmp(s, "\n")) {
+			/* Line feed/New line Mode */
+			if (cvt->lnm)
+				s = "\r\n";
 		}
 
-		mutexUnlock(cvt->lock);
-		mutexUnlock(ttypc->lock);
+		while (*s && !libtty_putchar(&cvt->tty, *s++, NULL))
+			;
 	}
+
+	mutexUnlock(cvt->lock);
+	mutexUnlock(ttypc->lock);
 }
 
 
-/* Waits for keyboard controller status bit with small timeout */
-static int ttypc_kbd_waitstatus(ttypc_t *ttypc, unsigned char bit, unsigned char state)
+static void ttypc_ps2_ctlthr(void *arg)
 {
-	unsigned int i;
+	ttypc_t *ttypc = (ttypc_t *)arg;
+	unsigned char dt;
+	unsigned char status;
 
-	for (i = 0; i < 0xffff; i++) {
-		if (!(inb((void *)((uintptr_t)ttypc->kbd + 4)) & ((1 << bit) ^ (state << bit))))
-			return EOK;
-		usleep(10);
+	mutexLock(ttypc->kmlock);
+	for (;;) {
+		status = ttypc_ps2_read_ctrl(ttypc);
+		if (OUTPUT_PENDING(status)) {
+			dt = inb((void *)ttypc->kbd);
```

Comment:
`ttypc_ps2_read()` is not needed, because the control register is already checked, right?

Reply:
Correct